x-pack/plugins/alerting/server/rules_client/lib/validate_attributes.ts

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { KueryNode } from '@kbn/es-query';
+import { get, isEmpty } from 'lodash';
+
+import mappings from '../../saved_objects/mappings.json';
+
+const astFunctionType = ['is', 'range', 'nested'];
+const mappingKeys = Object.keys(mappings);
+
+export const validateOperationOnAttributes = (
+  astFilter: KueryNode | null,
+  sortField: string | undefined,
+  searchFields: string[] | undefined,
+  excludedFieldNames: string[]
+) => {
+  if (sortField) {
+    validateSortField(sortField, excludedFieldNames);
+  }
+  if (!isEmpty(searchFields)) {
+    validateSearchFields(searchFields ?? [], excludedFieldNames);
+  }
+  if (astFilter) {
+    validateFilterKueryNode({ astFilter, excludedFieldNames });
+  }
+};
+
+export const validateSortField = (sortField: string, excludedFieldNames: string[]) => {
+  if (excludedFieldNames.filter((efn) => sortField.split('.')[0].includes(efn)).length > 0) {
+    throw new Error(`Sort is not supported on this field ${sortField}`);
+  }
+};
+
+export const validateSearchFields = (searchFields: string[], excludedFieldNames: string[]) => {
+  const excludedSearchFields = searchFields.filter(
+    (sf) => excludedFieldNames.filter((efn) => sf.split('.')[0].includes(efn)).length > 0
+  );
+  if (excludedSearchFields.length > 0) {
+    throw new Error(`Search is not supported on this field ${excludedSearchFields.join()}`);
+  }
+};
+
+interface ValidateFilterKueryNodeParams {
+  astFilter: KueryNode;
+  excludedFieldNames: string[];
+  hasNestedKey?: boolean;
+  nestedKeys?: string;
+  storeValue?: boolean;
+  path?: string;
+}
+
+export const validateFilterKueryNode = ({
```

Comment:
Personally, I find this recursion approach confusing since we need to add these additional trackers to help with the recursion.

I wonder if using like an iterative approach with a stack might work better. I tried it out and all tests pass but curious to your thoughts. I also simplified some checks because I think we can safely check for `alert.attributes.${excludedFieldName}` but maybe that assumption is bad?

```
export const validateFilterKueryNode = ({
  astFilter,
  excludedFieldNames,
}: ValidateFilterKueryNodeParams) => {
  const stack = [...astFilter.arguments];
  while (stack.length > 0) {
    const args = stack.shift();
    if (args.type === 'literal') {
      for (let x = 0; x < excludedFieldNames.length; x++) {
        if (
          typeof args.value === 'string' &&
          args.value.includes(`alert.attributes.${excludedFieldNames[x]}.`)
        ) {
          throw new Error(`Filter is not supported on this field ${args.value}`);
        }
      }
    }
    if (args.arguments) {
      stack.unshift(...args.arguments);
    }
  }
};
```

Reply:
I like the approach but I think we are going to miss some edge cases like nested fields and also I am preferring to re-use a techniques that is already working for validating kql filter on saved object -> https://github.com/elastic/kibana/blob/main/src/core/server/saved_objects/service/lib/filter_utils.ts#L99

Reply:
Sounds good! Let's keep it as is for now